File: android/measure/src/main/java/sh/measure/android/exporter/EventExporter.kt
package sh.measure.android.exporter

import androidx.annotation.VisibleForTesting
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.storage.Database
import sh.measure.android.storage.FileStorage
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Abstraction to run common functions for exporting of events and synchronizing different
 * exporters like [PeriodicEventExporter] and [ExceptionExporter].
 */
internal interface EventExporter {
    fun createBatch(sessionId: String? = null): BatchCreationResult?
    fun getExistingBatches(): LinkedHashMap<String, MutableList<String>>

    /**
     * Exports the events with the given [eventIds] in the batch with the given [batchId].
     *
     * @return the response of the export operation.
     */
    fun export(batchId: String, eventIds: List<String>): HttpResponse?
}

internal class EventExporterImpl(
    private val logger: Logger,
    private val database: Database,
    private val fileStorage: FileStorage,
    private val networkClient: NetworkClient,
    private val batchCreator: BatchCreator,
) : EventExporter {
    private companion object {
        const val MAX_EXISTING_BATCHES_TO_EXPORT = 5
    }

    @VisibleForTesting
    internal val batchIdsInTransit = CopyOnWriteArrayList<String>()

    override fun export(batchId: String, eventIds: List<String>): HttpResponse? {
        if (batchIdsInTransit.contains(batchId)) {
            logger.log(LogLevel.Warning, "Batch $batchId is already in transit, skipping export")
            return null
        }
        batchIdsInTransit.add(batchId)
        try {
            val events = database.getEventPackets(eventIds)
            if (events.isEmpty()) {
                // shouldn't happen, but just in case it does we'd like to know.
                logger.log(
                    LogLevel.Error,
                    "No events found for batch $batchId, invalid export request",
                )
                return null
            }
            val attachments = database.getAttachmentPackets(eventIds)
            val response = networkClient.execute(batchId, events, attachments)
            handleBatchProcessingResult(response, batchId, events, attachments)
            return response
        } finally {
            // always remove the batch from the list of batches in transit
            batchIdsInTransit.remove(batchId)
        }
    }

    override fun createBatch(sessionId: String?): BatchCreationResult? {
        return batchCreator.create(sessionId)
    }

    override fun getExistingBatches(): LinkedHashMap<String, MutableList<String>> {
        return database.getBatches(MAX_EXISTING_BATCHES_TO_EXPORT)
    }

    private fun handleBatchProcessingResult(
        response: HttpResponse,
        batchId: String,
        events: List<EventPacket>,
        attachments: List<AttachmentPacket>,
    ) {
        when (response) {
            is HttpResponse.Success -> {
                deleteEvents(events, attachments)
                logger.log(
                    LogLevel.Debug,
                    "Successfully sent batch $batchId",
                )
            }

            is HttpResponse.Error.ClientError -> {
                deleteEvents(events, attachments)
                logger.log(
                    LogLevel.Error,
                    "Client error while sending batch $batchId, dropping the batch",
                )
            }

            else -> {
                logger.log(LogLevel.Error, "Failed to send batch $batchId")
            }
        }
    }

    private fun deleteEvents(
        events: List<EventPacket>,
        attachments: List<AttachmentPacket>,
    ) {
        val eventIds = events.map { it.eventId }
        database.deleteEvents(eventIds)
        fileStorage.deleteEventsIfExist(eventIds, attachments.map { it.id })
    }
}

File: ios/Sources/MeasureSDK/Swift/Exporter/EventExporter.swift
//
//  EventExporter.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 17/10/24.
//

import Foundation

protocol EventExporter {
    func createBatch(_ sessionId: String?) -> BatchCreationResult?
    func getExistingBatches() -> [BatchEntity]
    func export(batchId: String, eventIds: [String]) -> HttpResponse?
}

final class BaseEventExporter: EventExporter {
    private let logger: Logger
    private let networkClient: NetworkClient
    private let batchCreator: BatchCreator
    private let batchStore: BatchStore
    private let eventStore: EventStore

    private let maxExistingBatchesToExport = 5
    private var batchIdsInTransit = Set<String>()

    init(logger: Logger, networkClient: NetworkClient, batchCreator: BatchCreator, batchStore: BatchStore, eventStore: EventStore) {
        self.logger = logger
        self.networkClient = networkClient
        self.batchCreator = batchCreator
        self.batchStore = batchStore
        self.eventStore = eventStore
    }

    func createBatch(_ sessionId: String?) -> BatchCreationResult? {
        return batchCreator.create(sessionId: sessionId)
    }

    func getExistingBatches() -> [BatchEntity] {
        return batchStore.getBatches(maxExistingBatchesToExport)
    }

    func export(batchId: String, eventIds: [String]) -> HttpResponse? {
        guard !batchIdsInTransit.contains(batchId) else {
            logger.internalLog(level: .warning, message: "Batch \(batchId) is already in transit, skipping export", error: nil, data: nil)
            return nil
        }

        batchIdsInTransit.insert(batchId)
        defer { batchIdsInTransit.remove(batchId) }

        guard let events = eventStore.getEvents(eventIds: eventIds), !events.isEmpty else {
            logger.internalLog(level: .error, message: "No events found for batch \(batchId), invalid export request", error: nil, data: nil)
            return nil
        }

        let response = networkClient.execute(batchId: batchId, events: events)

        handleBatchProcessingResult(response: response, batchId: batchId, events: events)
        return response
    }

    private func handleBatchProcessingResult(response: HttpResponse, batchId: String, events: [EventEntity]) {
        switch response {
        case .success:
            deleteEvents(batchId: batchId, events: events)
            logger.internalLog(level: .debug, message: "Successfully sent batch \(batchId)", error: nil, data: nil)
        case .error(let errorType):
            switch errorType {
                case .clientError(_, _): // swiftlint:disable:this empty_enum_arguments
                deleteEvents(batchId: batchId, events: events)
                logger.internalLog(level: .error, message: "Client error while sending batch \(batchId), dropping the batch", error: nil, data: nil)
                default:
                    break
                }
        }
    }

    private func deleteEvents(batchId: String, events: [EventEntity]) {
        let eventIds = events.map { $0.id }
        eventStore.deleteEvents(eventIds: eventIds)
        batchStore.deleteBatch(batchId)
    }
}
